### PR TITLE
Feature/interrupt with algebra

### PR DESCRIPTION
**Pull request by pchlupacek**
@mpilquist @SystemFw there is the interrupt implemented with Algebra. 

The idea is pretty simple, I had it on mind couple of weeks, but always failed to implement up to until now :-). Essentially the idea is that scope that is first to be interrupted is caching it`s next step in case of interruption and using that step when the interruption occurs. 

Also as part of this - 
 - Interrupted exception is gone
- Scope now takes type `O`
- `++`, `repeat`, `flatMap` reverted to M10 implementations

The implementation is bit noisy (if you can find a way how to make it cleaner I will really appreciate). Going forward I am thinking of introducing new type `Compile[F, O]` that will Contain current Scope[F, O] to cleanup algebra and also allow different compilation implementations to be plugged in (i.e. debugging compiler, javascript-optimized compiler etc..)

### CODE REVIEW

**Comment by pchlupacek** [comment]
FWIW I think the travis failure is unrelated

**Comment by SystemFw** [comment]
I think some instability in the `retry` spec is unavoidable (is testing repetitions in time), let's see if the failure is indeed a random one

**Comment by pchlupacek** [comment]
@SystemFw I think the issue is that delays is mutable unguarded ListBuffer. I think that spec needs to be made stronger, as the delays are accesses potentially from multiple threads.

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 1]
```diff
@@ -2793,7 +2771,7 @@ object Stream {
           val runStep =
             Algebra
               .compileScope(
-                scope,
+                scope.asInstanceOf[fs2.internal.CompileScope[F, UO]], // todo: resolve cast
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/Stream.scala]: Can the TODO be resolved by specifying UO as type param to getScope (instead of Nothing)?

Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/Stream.scala]: Unfortunatelly the result of `unconsAsync` is `Pull[F, Nothing, AsyncPull[F, Option[(Segment[O, Unit], Stream[F, O])]]]` so I would need then typecase UO back to Nothing, at least that's what compiler is asking me to do. 
I have plan to address in #1056.

**Comment by pchlupacek** [comment]
@mpilquist @SystemFw pls let me know when you will be done reviewing this. I have ready `merge` pull request that uses new interruption, and can just make PR once this is merged.

**Comment by SystemFw** [comment]
I'm in airports all day today, so it's more likely going to be tomorrow.
However, feel free to go ahead with only Michael's approval, I'm still
catching up on the core interpreter stuff

**Pull request merged**